### DISCUSSION TITLE
Reading Json data from file using structure object.

### DISCUSSION THREAD

**Discussion by Jhonlothan** [nlohmann/json#3231] [Q&A] [answered]
Hi, I have some data like 
[{"Name" : "jhon" , "age" : "13" , "section" : "A"}, 
{"Name" : "Larry" , "age" : "15" , "section" : "B"}, 
{"Name" : "narin" , "age" : "12" , "section" : "A"}]
. 
. 
. 
.
Like this in my file.json file. How can I read this data using structure object?

**Comment by nlohmann** [comment]
This would be invalid JSON, as there is no surrounding array or commas between the objects, right?

**Reply by Jhonlothan** [reply]
Sorry for that... I updated it.

**Reply by nlohmann** [reply]
See https://json.nlohmann.me/api/basic_json/parse/ (examples https://json.nlohmann.me/api/basic_json/parse/#examples)

**Reply by Jhonlothan** [reply]
Thank you . One more thing is, if I have data like this 

{"Anna" : { 
      "age": 18,
      "profession": "student"},
 "Ben" : {
      "age" : "nineteen",
      "profession": "mechanic"}
 }


Can I access particular key like this 
cout<<people["Anna"]["profession"]; ?

**Reply by nlohmann** [reply]
Yes

**Reply by IssaAboudi** [reply]
> Thank you . One more thing is, if I have data like this
> 
> {"Anna" : { "age": 18, "profession": "student"}, "Ben" : { "age" : "nineteen", "profession": "mechanic"} }
> 
> Can I access particular key like this cout<<people["Anna"]["profession"]; ?

I wanted to ask how to create this result. I've put my items in a JSON array but I'm not sure how to give it the value before the opening curly brace - for example the "Anna" before the opening {
 
For my purposes I am populating the JSON from code so I can't just write it that way to begin with

EDIT:
--
NEVERMIND TO THE ABOVE

How would I check the first category without knowing the field. Say I didn't know the first field was "Anna", I wouldn't be able to search for entries under that first field. How do I find that from a file?

**Reply by nlohmann** [reply]
You can iterate objects keys/values with the items function, see https://json.nlohmann.me/api/basic_json/items/

**Reply by Jhonlothan** [reply]
Hi, this iteration of objects is fetching the keys and values in alphabetical order. Is there is possibility to fetch in the same order as in json file?

**Reply by nlohmann** [reply]
See https://json.nlohmann.me/features/object_order/

**Reply by Jhonlothan** [reply]
Thank you very much

**Reply by PreetM07** [reply]
Hi, 
Not pretty sure but how do I access:
Items > source > text of following object.
{"total":2627,"items":[{"id":144721,"prevSegmentId":null,"nextSegmentId":144722,"number":1,"order":1.0,"documentId":"xxx","subtitleId":[],"topicId":"6f3c2182-9e04-4db7-a8b9-41106ad19ca7","commentState":0,"localizationContext":["check"],"wordsCount":5,"isSplitting":false,"paragraphId":1,"targetTextLengthLimit":null,"canModifyTextLengthLimit":false,"segmentRepetitionState":0,"source":{"segmentId":144721,**_"text":"check"_**,"languageId":9,"tags":[],"placeholders":[]},"targets":[{"segmentId":144721,"languageId":71,"documentId":"xxx","text":"","lockType":0,"tags":[],"placeholders":[],"revisions":[],"stageNumber":1,"isConfirmed":false,"errors":[],"lockUserId":null,"workflowRollbackStep":null,"machineTranslations":[],"textForNonAlphabeticAutoFill":null,"termEntries":[],"insertResourceType":0,"matchPercentage":0}],"images":[],"previewUrl":null,"defaultFormatting":null,"lockType":0}]}

items() gives me only outer object i.e. "items"

**Reply by Jhonlothan** [reply]
If your json object is j, then cout<<j[items][source][text]; will give the required data.

**Reply by jhonabraham22** [reply]
> See https://json.nlohmann.me/api/basic_json/parse/ (examples https://json.nlohmann.me/api/basic_json/parse/#examples)

Hi in my case file contains individual lines in json format, but the entire file is not in json format. This is incoming stream of data so I can't change the file format. How can I read each line from file and parse using json ? Can you please help.

{"Name" : "jhon" , "age" : "13" , "section" : "A"}
{"Name" : "Larry" , "age" : "15" , "section" : "B"}
{"Name" : "narin" , "age" : "12" , "section" : "A"}

**Reply by nlohmann** [reply]
I will add an example later.

**Reply by Jhonlothan** [reply]
It will be helpful for me also... Waiting for your example.

**Reply by PreetM07** [reply]
> If your json object is j, then cout<<j[items][source][text]; will give the required data.

It gives type error while building. 
and if used strings as j["items"]["source"]["text"]...
it gives following runtime error
terminate called after throwing an instance of 'nlohmann::detail::type_error'
  what():  [json.exception.type_error.305] cannot use operator[] with a string argument with array

**Reply by nlohmann** [reply]
Here is a branch to add documentation and an example for JSON Lines (newline-separated JSON text): https://github.com/nlohmann/json/pull/3247/files. Please let me know if this is helpful or if you need more details.

**Comment by jhonabraham22** [comment, marked as answer]
Instead of taking input in code, Can you please directly read from a file. And can you please show how to access particular key value in any one of the json line?.

**Reply by nlohmann** [reply]
- Reading from a file is the same: instead of passing a `std::stringstream`, you pass a `std::ifstream`.
- Accessing values can be done with these functions: https://json.nlohmann.me/features/element_access/

**Reply by Jhonlothan** [reply]
Hi if my data is 
Hi, I have some data within a file like this, 

{"Name" : "jhon" , "age" : "13" , "section" : "A"}
{"School" : "Larry" , "city" : "NewYork"}
{"Name" : "narin" , "age" : "12" , "section" : "A"}
{"School" : "Larry" , "city" : "NewYork"}

Two kinds of json objects in two lines. Iam able to parse line by line and able to access values with a json object like  j[" city"] . It is printing value of particular key if it is there in the parsed line otherwise showing null value. 
But when I try to assign it to a class member like 
Obj.Classmember1 = std::string (j["city"]) ; 
It is assigning correctly when value of key is found in parsed line. But in case of null means if value of key is not present in parsed line then it's showing error.302 how should I overcome this? Kindly help with this.

**Reply by nlohmann** [reply]
You can use `j.value("city", "placeholder")` (see https://json.nlohmann.me/api/basic_json/value/) or first check with `j.contains("key")` (see https://json.nlohmann.me/api/basic_json/contains/).

**Reply by Jhonlothan** [reply]
Hi, 
I used j.value but it's throwing an error 302 type must be a string, but is null. I assigned default value is of same type as key value. 
Here is my assignment

Obj.Classmember1 = std::string (j.value("city", "Texas")) ;

**Reply by nlohmann** [reply]
Then use contains and is_string.

**Reply by Jhonlothan** [reply]
Can you please show how contains and is_string is used for the following class member assignment

Obj.Classmember1 = std::string (j.value("city", "Texas")) ;

**Reply by nlohmann** [reply]
```cpp
if (j.contains("city") && j["city"].is_string()) {
  Obj.Classmember1 = j["city"].get<std::string>();
}
```

**Reply by Jhonlothan** [reply]
Thank you very much.It worked . But i have lot of class members to assign. Do I need to use if loop for every class member?

**Comment by gregmarr** [comment]
This should only be required if you have `null`s in your JSON data, or if you are using `j["city"]` on a non-const `json` object.  This will create a `city: null` key/value pair.  If you can figure out where that is happening and eliminate it, then you can use `j.value("city", "Texas")`.  If your JSON data can have `"city": null`, then you have to go with `contains() and is_string()` but you can put that in a function and re-use it.

**Reply by Jhonlothan** [reply]
Hi, is there any way to convert nlohmann::basic_json<>::string_t to unit32_t? 
I want to read data from json file and assign it to a struct member of type unit32_t.

**Reply by nlohmann** [reply]
No, you have to do this conversion yourself, e.g. via `std:: stoul`.
